### PR TITLE
Internal: Improved 'minimum_installation_version' functionality to support beta/cloud [ED-12173]

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: core/experiments/manager.php [hunk 1]
```diff
@@ -155,6 +155,19 @@ public function add_feature( array $options ) {
...
+
+		end( $installs_history );
+		$cleaned_version = preg_replace( '/-(beta|cloud|dev)\d*$/', '', key( $installs_history ) );
+
+		return version_compare(
+			$cleaned_version,
+			$version,
+			'>='
+		);
+	}
```
Comment (StyleShit) [inline core/experiments/manager.php:169]: why don't you fix it in the upgrade manager?

File: core/experiments/manager.php [hunk 2]
```diff
@@ -155,6 +155,19 @@ public function add_feature( array $options ) {
 		return $experimental_data;
 	}
 
+	private function install_compare( $version ) {
+		$installs_history = Upgrade_Manager::get_installs_history();
+
+		end( $installs_history );
```
Comment (StyleShit) [inline core/experiments/manager.php:161]: does it keep the pointer on the last index after calling `Upgrade_Manager::get_installs_history()` again? (in other words - does `get_installs_history` return the same reference?)

i.e.:

```PHP
$installs_history = Upgrade_Manager::get_installs_history();
end( $installs_history ); // last

// ... 

$installs_history_2 = Upgrade_Manager::get_installs_history();
key($installs_history_2; // still last?
```

File: tests/phpunit/elementor/core/experiments/test-manager.php [hunk 3]
```diff
@@ -372,36 +390,101 @@ public function test_get_active_features() {
...
+			'name' => 'test_feature',
+			'default' => Experiments_Manager::STATE_INACTIVE,
+			'mutable' => false,
+			'new_site' => [
+				'default_active' => true,
+				'minimum_installation_version' => '3.1.0',
+			],
+		] );
+
+		$is_test_feature_active = $this->experiments->is_feature_active( 'test_feature' );
```
Comment (StyleShit) [inline tests/phpunit/elementor/core/experiments/test-manager.php:480]: maybe also one for inactive?
and maybe just use a data provider for all those tests

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: core/experiments/manager.php (hunk 1 above)
Comment (ManorHazazE) [reply core/experiments/manager.php:169]: Can't change it...

File: core/experiments/manager.php (hunk 2 above)
Comment (ManorHazazE) [reply core/experiments/manager.php:161]: No. The pointer is on the variable.

File: core/experiments/manager.php [hunk 4]
```diff
@@ -93,7 +93,7 @@ public function add_feature( array $options ) {
 		$new_site = $experimental_data['new_site'];
 
 		if ( $new_site['default_active'] || $new_site['always_active'] ) {
-			$is_new_installation = Upgrade_Manager::install_compare( $new_site['minimum_installation_version'], '>=' );
```
[the remark this answers is not in the document]
Comment (ManorHazazE) [reply core/experiments/manager.php:96]: why? It is a private function that is being used just for this specific case

File: tests/phpunit/elementor/core/experiments/test-manager.php [hunk 5]
```diff
@@ -372,36 +390,101 @@ public function test_get_active_features() {
...
 
-		$experiments = $this->experiments;
+		$is_test_feature_active = $this->experiments->is_feature_active( 'test_feature' );
 
-		$is_test_feature_active = $experiments->is_feature_active( 'test_feature' );
+		$this->assertTrue( $is_test_feature_active );
+	}
+
+
+	public function test_is_feature_active__new_site_should_be_active() {
```
[the remark this answers is not in the document]
Comment (ManorHazazE) [reply tests/phpunit/elementor/core/experiments/test-manager.php:424]: Removed

File: tests/phpunit/elementor/core/experiments/test-manager.php [hunk 6]
```diff
@@ -372,36 +390,101 @@ public function test_get_active_features() {
...
-			time() => '3.1.0',
+		update_option( Manager::INSTALLS_HISTORY_META, [
+			'3.1.0' => time(),
 		] );
 
-		$this->add_test_feature( [
+		$this->experiments->add_feature( [
+			'name' => 'test_feature',
 			'default' => Experiments_Manager::STATE_INACTIVE,
+			'mutable' => false,
```
[the remark this answers is not in the document]
Comment (ManorHazazE) [reply tests/phpunit/elementor/core/experiments/test-manager.php:411]: Removed

File: tests/phpunit/elementor/core/experiments/test-manager.php [hunk 7]
```diff
@@ -20,12 +20,30 @@ class Test_Manager extends Elementor_Test_Base {
 	 */
 	private $experiments;
 
+	/**
+	 * @var array The state of the experiments before the bootstrap file of the tests changes it.
```
[the remark this answers is not in the document]
Comment (ManorHazazE) [reply tests/phpunit/elementor/core/experiments/test-manager.php:24]: Yes Sir

File: tests/phpunit/elementor/core/experiments/test-manager.php [hunk 8]
```diff
@@ -20,12 +20,30 @@ class Test_Manager extends Elementor_Test_Base {
...
 		parent::setUp();
 
+		add_action( 'elementor/experiments/feature-registered', [ $this, 'set_experiments_state' ], 1, 2 );
+
+		add_action( 'elementor/experiments/feature-registered', [ $this, 'reset_experiments_state' ], 9999, 2 );
+
 		$this->experiments = new Experiments_Manager();
 	}
 
+	public function set_experiments_state( Experiments_Manager $experiments_manager, array $experimental_data ) {
```
[the remark this answers is not in the document]
Comment (ManorHazazE) [reply tests/phpunit/elementor/core/experiments/test-manager.php:38]: ok

File: tests/phpunit/elementor/core/experiments/test-manager.php [hunk 9]
```diff
@@ -20,12 +20,30 @@ class Test_Manager extends Elementor_Test_Base {
...
+	 * @var array The state of the experiments before the bootstrap file of the tests changes it.
+	 */
+	private $experiments_state = [];
+
 	public function setUp() {
 		parent::setUp();
 
+		add_action( 'elementor/experiments/feature-registered', [ $this, 'set_experiments_state' ], 1, 2 );
+
+		add_action( 'elementor/experiments/feature-registered', [ $this, 'reset_experiments_state' ], 9999, 2 );
```
[the remark this answers is not in the document]
Comment (ManorHazazE) [reply tests/phpunit/elementor/core/experiments/test-manager.php:33]: Don't worry be חappy

File: tests/phpunit/elementor/core/experiments/test-manager.php [hunk 10]
```diff
@@ -372,36 +390,101 @@ public function test_get_active_features() {
...
+			'default' => Experiments_Manager::STATE_ACTIVE,
+		] );
 
 		$is_test_feature_active = $this->experiments->is_feature_active( 'test_feature' );
 
 		$this->assertTrue( $is_test_feature_active );
 	}
 
 	public function test_is_feature_active__new_site() {
-		update_option( Manager::get_install_history_meta(), [
```
[the remark this answers is not in the document]
Comment (ManorHazazE) [reply tests/phpunit/elementor/core/experiments/test-manager.php:383]: Yes boss

File: tests/phpunit/elementor/core/experiments/test-manager.php (hunk 3 above)
Comment (ManorHazazE) [reply tests/phpunit/elementor/core/experiments/test-manager.php:480]: OK

**Review by StyleShit — DISMISSED** [review]

**Review by StyleShit — DISMISSED** [review]

**Review by StyleShit — APPROVED** [review]

**Pull request merged**